### DISCUSSION TITLE
Bulk Operations with personal licence

### DISCUSSION THREAD

**Discussion by BluePython339** [Vector35/binaryninja-api#2115] [Q&A] [answered]
Hey, how would i go about performing bulk operations of multiple binary files and extracting say the exact assembly from a binary and write it to a text file? I noted the python console in the GUI but that isn't really practical for doing these bulk operations

**Comment by psifertex** [comment, marked as answer]
That's the only mechanism supported. The "headless" feature as described on the purchase page explains the differences between the commercial product and the non-commercial license, and this is exactly it: https://binary.ninja/purchase/

That said, there's no reason you can't write plugins that do bulk processing as long as the UI is running. 

For example, drop the following file into your [plugin folder](https://docs.binary.ninja/guide/plugins.html#using-plugins):

```
#batch.py
from binaryninja import open_view
from glob import glob
def go():
    for bin in glob("/bin/*"):
        with open_view(bin, update_analysis=False) as bv:
            print(f"Opening {bv.file.filename} which has {len(bv.functions)} functions")
```

Then run it via the console with:

```
import batch
from importlib import reload
reload(batch);batch.go()
```

Then you can just `<up><enter>` in the console to reload any changes from the file and re-run it. 

I realize it's more convenient to be able to just directly run python stand-alone straight into the core, but that's one of the very few limitations we have on the non-commercial version versus commercial. Note that the student discount also applies to the commercial version, so many students still chose to get the commercial version for just this capability.

**Reply by psifertex** [reply]
As for your specific example, I modified the example above to dump the disassembly for all files in a folder. Note that I'm making analysis run faster but not fully analyzing the files with the `update_analysis=False` below. 

```
#batch.py
from binaryninja import open_view
from glob import glob
import os
def go():
    for bin in glob("/bin/*"):
        with open_view(bin, update_analysis=False) as bv:
            output_name = os.path.join("/tmp", os.path.basename(bin) + ".txt")
            with open(output_name, "w+") as f:
                f.write("\n".join(map(str, bv.linear_disassembly)))

```

**Reply by tosanjay** [reply]
Hi @psifertex 
In the same setup, suppose that I want to analyze the binary that is already open in UI (with a script that I have written myself, like batch, in your example). This means that I will have to get binary_view OR open_view for the current loaded binary. I am not able to find a way to get this. Can you please provide some pointer/solution to that?
Note: I am just beginning to learn BN (1 day old into this) and I can't use headless mode.

Thanks & Regards
-Sanjay

**Reply by xusheng6** [reply]
If a binary is already open in the UI, `bv` is the binary view you need. It is a magical variable we injected to the Python environment.

**Reply by tosanjay** [reply]
PS: I found "binaryninja.interaction" to ask for the file I want to analyze, but it is still entering the file name explicitly (again). If no solution to my original query, this is a workaround.

**Reply by tosanjay** [reply]
> If a binary is already open in the UI, `bv` is the binary view you need. It is a magical variable we injected to the Python environment.

But this is available via python console. What I am looking for is to have my module (analysis script) loaded (import my_script), which does some analysis on the opened binary.

**Reply by psifertex** [reply]
Registering your plugin with: https://api.binary.ninja/binaryninja.plugin-module.html?highlight=register#binaryninja.plugin.PluginCommand.register

will get you a callback with the current BV attached. Then you trigger it via the tools/plugin menu or the command-palette. 

You can also use the UI Plugin APIs to walk all open tabs for more complicated interactions but I don't recommend that route unless it's necessary for other reasons.

**Reply by tosanjay** [reply]
Thank you @psifertex. this seems very nice. I will try this method and report back :)

**Reply by tosanjay** [reply]
Hi @psifertex, Yes, your solution works very well. Thank you :)
